### PR TITLE
Add pipelines::masked_language and codebert support

### PR DESCRIPTION
**Pull request by Vincent-Xiao**
1. Add support for loading local model in SequenceClassificationConfig:
   Somtimes, the https://huggingface.co/models doesn't contains a rust model (eg. rust_model.ot) of some bert. Hence, it's 
   necessary to load local model easily in SequenceClassificationConfig.

2.  Add piplines::masked_language for masking texts more easily; add two examples (masked_language_roberta.rs and masked_language.rs) for utilizing the piplines::masked_language.

3.  Add support  for codebert; add an example (codebert_example.rs) including on feature_extraction, masked_language and sequence_classification; also upload rust models (rust_model.ot) of huggingface/CodeBERTa-language-id (merged) and microsoft/codebert-base-mlm (waiting to be merged) to  https://huggingface.co/models:
   Rust-bert currently support codebert by a roberta way. Adding codebert support all alone in src/codebert helps inferencing codebert applications more convenient and also is useful for dealing with the open issue (https://github.com/guillaume-be/rust-bert/issues/219).

### CODE REVIEW

File: src/pipelines/masked_language.rs [hunk 1]
```diff
@@ -0,0 +1,581 @@
...
+        var_store.load(weights_path)?;
+        Ok(MaskedLanguageModel {
+            tokenizer,
+            language_encode,
+            var_store,
+            max_length,
+        })
+    }
+
+    fn prepare_for_model<'a, S>(&self, input: S) -> Tensor
```
Comment (guillaume-be) [inline src/pipelines/masked_language.rs:447]: I believe this should set the attention mask and mask the `mask_token` positions as well (probably return a tuple of tensors)

File: examples/codebert_example.rs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    //    Run model
+    let output = sequence_classification_model.predict(&input);
+    for label in output {
+        println!("{:?}", label);
+    }
+
+    Ok(())
+}
+
+fn main() {
```
Comment (guillaume-be) [inline examples/codebert_example.rs:116]: Can you please ensure we only run relevant example (providing valid outputs)? General tests for model functionality can be moved to integration tests

File: examples/masked_language_roberta.rs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+// limitations under the License.
+
+extern crate anyhow;
+use rust_bert::pipelines::common::ModelType;
+use rust_bert::pipelines::masked_language::{MaskedLanguageConfig, MaskedLanguageModel};
+use rust_bert::resources::RemoteResource;
+use rust_bert::roberta::{
+    RobertaConfigResources, RobertaMergesResources, RobertaModelResources, RobertaVocabResources,
+};
+fn main() -> anyhow::Result<()> {
```
Comment (guillaume-be) [inline examples/masked_language_roberta.rs:20]: If the changes for the `mask_token` config suggestion (see below) are implemented, it would be nice to keep only 1 example with RoBERTa and use `[MASK]` as a mask token to illustrate capabilities and ensure cotrrectness

File: examples/sentiment_analysis_fnet.rs [hunk 4]
```diff
@@ -31,7 +31,8 @@ fn main() -> anyhow::Result<()> {
 
     let sentiment_config = SentimentConfig {
         model_type: ModelType::FNet,
-        model_resource,
+        model_resource: Some(model_resource),
+        model_local_resource: None,
```
Comment (guillaume-be) [inline examples/sentiment_analysis_fnet.rs:35]: This can probably be reverted, see below

File: src/codebert/codebert_model.rs [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+        }
+    }
+}
+pub struct CodeBertFeatureExtractionOutput {
+    /// Hidden states for all intermediate layers
+    pub hidden_states: Tensor,
+    /// Attention weights for all intermediate layers
+    pub all_attentions: Option<Vec<Tensor>>,
+}
+pub struct CodeBertForFeatureExtractionModel {
```
Comment (guillaume-be) [inline src/codebert/codebert_model.rs:181]: Looks like this code corresponds to a pipeline and not a model. We now have `SentenceEmbeddings` in this library, please refer to the relevant pipeline `sentence_embeddings`. Bert-like and RoBERTa-like models can be readily implemented from their masked-model implementation.

I believe you could just declare an alias, similar to https://github.com/guillaume-be/rust-bert/blob/59c0e668d68eb80c674100049ecd37c35c9ab4b5/src/roberta/roberta_model.rs#L997

File: src/codebert/codebert_model.rs [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+        "https://huggingface.co/huggingface/CodeBERTa-language-id/resolve/main/merges.txt",
+    );
+    /// Shared under Apache 2.0 license by the Microsoft team at <https://github.com/microsoft/CodeBERT>. Modified with conversion to C-array format.
+    pub const CODEBERT_MLM: (&'static str, &'static str) = (
+        "codebert-mlm/model",
+        "https://huggingface.co/microsoft/codebert-base-mlm/resolve/main/merges.txt",
+    );
+}
+
+pub struct CodeBertForFeatureExtraction {
```
Comment (guillaume-be) [inline src/codebert/codebert_model.rs:123]: I believe you could declare an alias as in https://github.com/guillaume-be/rust-bert/blob/59c0e668d68eb80c674100049ecd37c35c9ab4b5/src/roberta/roberta_model.rs#L997

File: src/codebert/codebert_model.rs [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+        "https://huggingface.co/huggingface/CodeBERTa-language-id/resolve/main/rust_model.ot",
+    );
+    /// Shared under Apache 2.0 license by the Microsoft team at <https://github.com/microsoft/CodeBERT>. Modified with conversion to C-array format.
+    pub const CODEBERT_MLM: (&'static str, &'static str) = (
+        "codebert-mlm/model",
+        "https://huggingface.co/microsoft/codebert-base-mlm/resolve/main/rust_model.ot",
+    );
+}
+
+impl CodeBertConfigResources {
```
Comment (guillaume-be) [inline src/codebert/codebert_model.rs:69]: The easiest option would be to add these additional resources to https://github.com/guillaume-be/rust-bert/blob/59c0e668d68eb80c674100049ecd37c35c9ab4b5/src/roberta/roberta_model.rs#L35 - it seems there are no architectural differences between CodeBERT and RoBERTa.

File: src/codebert/mod.rs [hunk 8]
```diff
@@ -0,0 +1,72 @@
+//! # RoBERTa: A Robustly Optimized BERT Pretraining Approach (Liu et al.)
```
Comment (guillaume-be) [inline src/codebert/mod.rs:1]: I am unsure this module is required - tokenization and modeling seem to be identical for RoBERTa and CodeBERT

**Review by guillaume-be — COMMENTED** [review]
Hello @Vincent-Xiao 

Thank you for the pull request, and for creating Rust-version of the pretrained CodeBERT models. A few comments on my side:
1. Local model loading is already possible, the following illustrate how one would point to a local file:
```rust
    let config = SequenceClassificationConfig {
        model_type: ModelType::Roberta,
        model_resource: Box::new(LocalResource {
            local_path: "path/to/CodeBERTa-language-id/rust_model.ot".into(),
        }),
        config_resource: Box::new(LocalResource {
            local_path: "path/to/CodeBERTa-language-id/config.json".into(),
        }),
        vocab_resource: Box::new(LocalResource {
            local_path: "path/to/CodeBERTa-language-id/vocab.json".into(),
        }),
        merges_resource: Some(Box::new(LocalResource {
            local_path: "path/to/CodeBERTa-language-id/merges.txt".into(),
        })),
        lower_case: false,
        strip_accents: None,
        add_prefix_space: None,
        device: Device::cuda_if_available(),
    };
```
2. Thank you for the suggestion of a masked language pipeline. I suggest adding an extra configuration option to facilitate masking by users (make the mask token model-agnostic)
3. The CodeBERT model seems to be identical to RoBERTa (model and tokenization) -- could you please advise if that is the case? If so, it would be much simpler to register the additional pre-trained models as RoBERTa variants

File: examples/codebert_example.rs [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+        )),
+        true,
+        None,
+        false,
+    );
+    //    Set-up model
+    let model = MaskedLanguageModel::new(config)?;
+    //    Define input
+    let input = [
+        "Looks like one <mask> is missing!",
```
[the remark this answers is not in the document]
Comment (Vincent-Xiao) [reply examples/codebert_example.rs:73]: > 

Hello@guillaume-be
Thanks for your review.
1. I tried to load local model as your suggestion, but error ocuurs: "the trait `ResourceProvider` is not implemented for `Box<LocalResource>`" and "expected struct `Box`, found struct `RemoteResource`", as described in following image.
image
image

2. I will add an extra configuration option for pipelines::masked_language as you suggested.
3. The codebert has an identical  model to roberta. However, some models in huggingface.co has a different prefix as roberta. For example, the model structure of roberta always is like "roberta.encoder.layer....."; some codebert model in  huggingface.co is "encoder.layer"; some is "roberta.encoder.layer"; some is "codebert.encoder.layer". Error would occurs while loading codebert model like a roberta way.Hence, supporting codebert in a seperate way may be useful to tackle the prefix difference.

[the remark this answers is not in the document]
Comment (Vincent-Xiao) [reply examples/codebert_example.rs:73]: Hello @guillaume-be Could you give me some advice on the first and third question? I am going to finish this PR as soon as possible!

[the remark this answers is not in the document]
Comment (guillaume-be) [reply examples/codebert_example.rs:73]: Hello @Vincent-Xiao ,

Regarding 1. the error occurs because you are trying to mix `LocalResource` with `RemoteResource`. While the models can take either of them (pointing to a remote model or local model files), it is not possible to mix both. In your example you would have to update the config and tokenizer files to also point to local files.

Rgarding 3., could you give an example of a model that uses a different prefix for the model (`codebert.encoder.layer`)? I'd be interested to see how the Python version fixes this problem as the Python's implementation of transformers does not include code specific to `codebert`

File: src/pipelines/masked_language.rs [hunk 10]
```diff
@@ -0,0 +1,581 @@
...
+    pub merges_resource: Option<Box<dyn ResourceProvider + Send>>,
+    /// Automatically lower case all input upon tokenization (assumes a lower-cased model)
+    pub lower_case: bool,
+    /// Flag indicating if the tokenizer should strip accents (normalization). Only used for BERT / ALBERT models
+    pub strip_accents: Option<bool>,
+    /// Flag indicating if the tokenizer should add a white space before each tokenized input (needed for some Roberta models)
+    pub add_prefix_space: Option<bool>,
+    /// Device to place the model on (default: CUDA/GPU when available)
+    pub device: Device,
+}
```
[the remark this answers is not in the document]
Comment (Vincent-Xiao) [reply src/pipelines/masked_language.rs:99]: How to get the tokenizer's mask token ? For example, the BertTokenizer in rust_tokenizers does not have a "mask_token" member and  related functions returning "mask_token". If there is not a straightforward way to get mask_token, maybe I need to predefine "mask_token" for various  tokenizers and add it to the TokenizerOption.

File: examples/codebert_example.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (Vincent-Xiao) [reply examples/codebert_example.rs:73]: Hello @guillaume-be 
1. My implemented method for loading local model in this PR supports mix LocalResource with RemoteResource. I think supporting this mix way still means something. The huggingface.co always have config files but may not contains a rust model. However, if you don't think it's worth supporting this, I will withdraw the sequenceConfig update.
3. The prefix of microsoft/codebert-base is "encoder.layer" without "roberta.".  And our team trained a codebert-based model (Vincent-Xiao/codebert-curs) to classify the "unsafe" keywords in rust program. This mode has a "codebert.encoder.layer" prefix.

File: src/pipelines/masked_language.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (Vincent-Xiao) [reply src/pipelines/masked_language.rs:101]: @guillaume-be 
I have added an optional mask_token String field for masked_language pipeline. It support regex replace when mask_token in input sentences is not matched to current model.

File: examples/codebert_example.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (Vincent-Xiao) [reply examples/codebert_example.rs]: @guillaume-be 
1. There are an common scenario where huggingface.co/models don't have a rust_mode.ot of a specific bert . In such circumstance, if someone want to deploy this specific model in rust-bert, user must convert the pytorch_model.bin to rust_model.ot offline. And user may load local model files and remote config files (such as config.json or merges.txt from huggingface.co/models). Therefore, it's necessary to support the mixed way (pointing to a remote config and local model files). My implemented LanguageConfig support this scenario well.

3. I have withdrawn the codebert support in consideration of your suggestion.

[the remark this answers is not in the document]
Comment (guillaume-be) [reply examples/codebert_example.rs]: Hello @Vincent-Xiao ,

For 1. there is a way to implement such functionality without having to duplicate field. The issue is that the generic `<R>` is shared between all resource types. I am pushing a pull request hat relaxes this restriction, see https://github.com/guillaume-be/rust-bert/pull/291. Can you please update the style of your proposal to match the linked pull request? This should eliminate the need for a duplicate `local_model_resource`.

The following works on the proposed implementation:
```rust
extern crate anyhow;

use rust_bert::pipelines::common::ModelType;
use rust_bert::pipelines::sequence_classification::{
    SequenceClassificationConfig, SequenceClassificationModel,
};
use rust_bert::resources::{LocalResource, RemoteResource};
use tch::Device;

fn main() -> anyhow::Result<()> {
    //    Set-up model
    let config = SequenceClassificationConfig {
        model_type: ModelType::Roberta,
        model_resource: Box::new(LocalResource {
            local_path: "path/to/CodeBERTa-language-id/rust_model.ot".to_string(),
        }),
        config_resource: Box::new(RemoteResource {
            url:
                "https://huggingface.co/huggingface/CodeBERTa-language-id/resolve/main/config.json"
                    .to_string(),
            cache_subdir: "codeberta/config".to_string(),
        }),
        vocab_resource: Box::new(RemoteResource {
            url: "https://huggingface.co/huggingface/CodeBERTa-language-id/resolve/main/vocab.json"
                .to_string(),
            cache_subdir: "codeberta/vocab".to_string(),
        }),
        merges_resource: Some(Box::new(RemoteResource {
            url: "https://huggingface.co/huggingface/CodeBERTa-language-id/resolve/main/merges.txt"
                .to_string(),
            cache_subdir: "codeberta/merges".to_string(),
        })),
        lower_case: false,
        strip_accents: None,
        add_prefix_space: None,
        device: Device::cuda_if_available(),
    };

    let sequence_classification_model = SequenceClassificationModel::new(config)?;

    //    Define input
    let input = [
        "
        def f(x):
            return x**2
        ",
        "const foo = 'bar'",
        "outcome := rand.Intn(6) + 1",
    ];

    //    Run model
    let output = sequence_classification_model.predict(&input);
    for label in output {
        println!("{:?}", label);
    }

    Ok(())
}
```

**Comment by guillaume-be** [comment]
@Vincent-Xiao would you be OK for me to help with this pull request and push a few of the changes requested? It'd be great to have this model and feature integrated in the library

**Comment by Vincent-Xiao** [comment]
> @Vincent-Xiao would you be OK for me to help with this pull request and push a few of the changes requested? It'd be great to have this model and feature integrated in the library
@guillaume-be I am so busy during this period and very sorry for not up-to-date with this repo. Do you mean the conflicting files? If yes, I will fix this within one week. Otherwise, please tell me the detail that I can do to help with this PR.

File: examples/codebert_example.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (Vincent-Xiao) [reply examples/codebert_example.rs]: @guillaume-be Thanks for your suggestion. I have revoked this functiuon and updated the the implemnetation for local moddel support as your suggesting way.

**Comment by guillaume-be** [comment]
Thank you again for the PR!
